### ISSUE TITLE
Firefox Snap & KeePassXC AppImage does not work with Ubuntu

### ISSUE THREAD

**Issue by HDInfautre** [keepassxreboot/keepassxc-browser#1828]
hello
I have set up Keepass correctly according to https://keepassxc.org/docs/KeePassXC_GettingStarted.html#_setup_browser_integration

Everything works fine with brave.
It does not work with Firefox 108.2

I removed  (then reinstalled)  the browser addon and re-run the keepassxc-snap-helper.sh script then re-run the browsers.

When I click on the KeepassXC icon, nothing happens
No databases connected in the browser settings page
the KeepassXC database is open and has the focus

## Debug info
[Capture d’écran du 2023-01-18 12-44-37](https://user-images.githubusercontent.com/6817936/213163546-19da1315-07eb-4c79-adab-6df40a60ccc5.png)
I don't see the little green button next to the input.
KeePassXC appImage- 2.7.4
KeePassXC-Browser - 1.8.4
Operating system:Ubuntu 22.10
Browser:  Firefox 109 nOK !! - Brave Version 1.46.153 : Ok

**Comment by varjolintu** [comment]
Probably duplicate of: https://github.com/keepassxreboot/keepassxc-browser/issues/1426. Check your Snap stuff.

**Comment by HDInfautre** [comment]
Hi
Ok no worries about coming back here if you read.
Did you understand that **it worked for a few days and that the only change comes after the change from 108.2 to 109**
That's why I think it's the same problem (and therefore not a spam), it's the 1st time **I don't even see the keepass icon next to the input fields**
If not, please tell me what I should do precisely? or wait for a new version for Firefox 109?

**Comment by varjolintu** [comment]
I don't know the details here. Is the Firefox installed as Snap (it's default in Ubuntu). Ubuntu just recently fixed Native Messaging (the API that the extension uses for message transferring) for Snap Firefox, so they've probably broken it?

Things you can try:
- Read the issue thread I linked earlier and see if it helps
- Read https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide
- Uninstall Snap Firefox and just use the PPA version

**Comment by HDInfautre** [comment]
I have re-read the thread, but I don't see anything helpful or too technical.
I had the same problem as some in previous years when deleting snap. i feel like keeping it.
Here is the feedback
If I understand correctly, this would be an ubuntu problem?
Can't you do anything?

I have re-read the thread, but I don't see anything helpful or too technical.
I had the same problem as some in previous years when deleting snap. i feel like keeping it.
Here is the feedback:
flatpak permissions webextensions
Table         Object                          App          Permissions Data
webextensions org.keepassxc.keepassxc_browser snap.firefox yes         0x00
webextensions mozillavpn                      snap.firefox yes         0x00
It looks good to me.

**Comment by varjolintu** [comment]
If the connection just broke when Ubuntu updates its Firefox, it's very probable that it's Snap's fault. Can you try to downgrade?

**Comment by HDInfautre** [comment]
HI
I think I understood. I've tried a lot of different configurations.
The only one that works on ubuntu 22.10 is Firefox (snap) with KeepassXC (snap)!

I'd like to run Firefox (snap) with KeepassXC (appimage) to have a portable version on a key!
I hope this helps to solve !?

Yes I can make suggestions, I have another one ;-)

**Comment by varjolintu** [comment]
Does the Native Messaging scripts in Snap Firefox point to your AppImage?

**Comment by HDInfautre** [comment]
Hi
I don't know, I'm not a ubuntu geek. 
For clarification (which may block).
I had seen, with another person here a long time ago, putting the configuration in xxx.home to make a portable application (see picture)
Can you tell me the simple check/change to do on this system?
![Capture d’écran du 2023-01-20 14-32-41](https://user-images.githubusercontent.com/6817936/213707032-53417657-666d-4b77-a721-f1f324e1a046.png)

**Comment by HDInfautre** [comment]
Hi
It works with firefox/snap and keepassxc/snap
But, I would prefer keepassxc/ keepassxc/snap
Were you able to reproduce the problem with an appImage?
Or if so is it possible to fix?

**Comment by HDInfautre** [comment]
Hi
This morning, after a week of correct operation, it is again impossible to connect keepassbrowser (firefox v109) with keepassXC. the computer was switched off at the weekend.
I have restarted the keepassxc-snap-helper.sh program without success.
Would you mind reopening the ticket?
Do you have a simple test to move you towards a solution?

**Comment by varjolintu** [comment]
The Snap helper script only helps if you are using KeePassXC as a Snap. I'll have to test this myself when I have the time.

**Comment by HDInfautre** [comment]
Ok I'm waiting and hoping for a more stable version under Ubuntu 22.10 for firefox/snap + keepassXC/snap  
Thanks for looking at firefox/snap + keepassXC/appimage as I loved this possibility to have a portable version of keepassXC?
but can you reopen this ticket for public follow up? 
Thank you
regards

**Comment by varjolintu** [comment]
I can, with a modified topic.

**Comment by varjolintu** [comment]
Both KeePassXC Flatpak and Appimage are having the same problem with Firefox Snap: The browser is trying to start the `keepassxc` process inside the image instead of `keepassxc-proxy`. So maybe the Firefox Snap is starting something differently with Native Messaging and KeePassXC cannot identify if the process is trying to start the proxy or the main application.

**Comment by HDInfautre** [comment]
Hello
Thank you for trying.
I had already followed the many messages about the container aspect (appimage) and the difficulty of "messaging" :-(
I thought some progress had been made. 
I guess you couldn't do much about it.
Nevertheless, have you reported this to the responsible developers? ubuntu? what are the perspectives?

**Comment by benbucksch** [comment]
Possible solutions, to fix this:
* Let KeyPassXC (Linux app, with user permissions) modify the AppArmor rules to allow the connection
* Work with Ubuntu Firefox snaps maintainer to allow the connection
* In the meantime: Instead of an error message in the addon, detect the problem and show clear instructions what the user should do. (This is not a fix for this bug, but only a hotfix, a temporary workaround until it's fixed properly.)

**Comment by benbucksch** [comment]
> isn't possible

Care to elaborate? Are the Ubuntu Firefox snap maintainers not reachable? It is technically impossible? (I doubt that.)

> Instead of an error message in the addon, detect the problem and show clear instructions what the user should do

You can definitely do that, and that would be the very minimum to do.

Have you considered using other communication mechanisms between addon and app?

As is, it looks like a bug in KeePass addon to the end user.

**Comment by fossil-free** [comment]
firefox snap and keepassxc worked for me in Ubuntu 22.04 after initialising webextensions.
Addon integration broke now when upgrading to 24.04. Therefore #2370 doesn't seem like a duplicate - may those be different issues? Maybe some path for webextensions changed?

--
Firefox 133.0b2 64bit
Firefox snap for Ubuntu canonical-002 - 1.0

KeePassXC - Version 2.7.9
Revision: 8f6dd13
Qt 5.15.13
OS: Ubuntu 24.04.1 LTS
CPU architecture: x86_64
kernel: linux 6.8.0-48-generic

Activated extensions:
- Auto-Type
- Browser-Integration
- Passkeys
- SSH-Agent
- KeeShare
- YubiKey
- Secret-Service-Integration

Kryptographische Bibliotheken:
- Botan 2.19.3

**Comment by varjolintu** [comment]
@fossil-free This is probably purely Ubuntu/Snap isse.

**Comment by fossil-free** [comment]
Firefox snap's console is giving following error when trying to connect to the keepassxc database by pushing the respective buttons:
"File at path "/usr/bin/keepassxc-proxy" does not exist, or is not executable"
keepassxc-proxy is existent though, and also flagged as executable. Might the file permissions make a difference here? Though root:root seems legit to me for bin files.

keepassxc-proxy starts several times a second with a new PID each time, stops immediately though. Weird that it's starting via autoconnect while saying it's not existent when using the manual method. 

`MOZ_LOG=NativeMessagingPortal:5 snap run firefox` returns:
`D/NativeMessagingPortal will not be used`

configuration path correctly points to my current snap:
~/snap/firefox/common/.mozilla/firefox/n80h6q6h.default/native-messaging-hosts/

To me it does seem like a keepassxc-proxy connection problem.
Any ideas how to fix it?

(Besides, I am somewhat confused by the existence of the apt package webext-keepassxc-browser found in the docs. Installing it didn't make any difference though. Why does it exist in parallel to the firefox and chromium addons?)

**Comment by droidmonkey** [comment]
This is 100% on the Firefox Snap, please file a bug with them.

**Comment by fossil-free** [comment]
I filed a mozilla bug now, thanks for pointing me there:
https://bugzilla.mozilla.org/show_bug.cgi?id=1930119

The snap profile configuration folder was actually set automatically to the snap directory (I just checked it via the advanced tab). I tried the ~/.mozilla directory as well and will keep it in mind checking both options.

In the keepassxc-browser [troubleshooting guide](
https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide) the flatpak permission point is mentioned as necessary in 5) Ubuntu snap... and indeed this had been necessary for me to get keepassxc-browser working in Ubuntu 22.04.

**Comment by fossil-free** [comment]
The firefox bug exists in snap firefox 133 + 134.

It can be solved by this workaround until the patch goes live in 135 or so:
1. visit about:config in the URL bar
2. search for "widget.use-xdg-desktop-portal.native-messaging" and set it from "0" to "2"

My keepass works flawlessly again with this hotfix.

**Comment by droidmonkey** [comment]
> Firefox 133.0b2 64bit

This is the beta channel for Firefox, the current stable release for snap is 132.0.1-2. I also saw the bug in 132.

**Comment by varjolintu** [comment]
KeePassXC Snap package should now work with the browser extension:
https://www.omgubuntu.co.uk/2025/02/keepassxc-snap-update-adds-native-messaging-support-i-e-browser-integration

**Comment by boukendesho** [comment]
> KeePassXC Snap package should now work with the browser extension: https://www.omgubuntu.co.uk/2025/02/keepassxc-snap-update-adds-native-messaging-support-i-e-browser-integration

For anyone seen this, this is not true if you are running snap chromium or any chromium-based snap browser. It only works with snap firefox.
